src/foam/nanos/crunch/connection/CapabilityPayloadDAO.js

```diff
@@ -50,6 +55,52 @@ foam.CLASS({
             setX(x);
             setDelegate(delegate);
           } 
+
+          private Map<String, FObject> walkGrantPath(List grantPath, X x) {
+            DAO userCapabilityJunctionDAO = (DAO) x.get("userCapabilityJunctionDAO");
+            var capabilityDAO = (DAO) x.get("capabilityDAO");
+            var user = ((Subject)x.get("subject")).getUser();
+            var businessId = user.getId();
+            var userId = user.getCreatedBy();
+
+            // Collect all capabilities that belong to either the user or business
+            var capabilities = new ArrayList<Capability>();
+            ((ArraySink) userCapabilityJunctionDAO.where(
+              OR(
+                EQ(UserCapabilityJunction.SOURCE_ID, userId),
+                EQ(UserCapabilityJunction.SOURCE_ID, businessId)
+              )
+            ).select(new ArraySink())).getArray().forEach((ucj) -> {
+              var capability = (Capability) capabilityDAO.find(((UserCapabilityJunction)ucj).getTargetId());
+              if ( capability != null )
+                capabilities.add(capability);
+            });
+      
+            Map<String,FObject> capabilityDataObjects = new HashMap<>();
+      
+            for ( Object item : grantPath ) {
+              // TODO: Verify that MinMaxCapability lists can be ignored here
+              if ( item instanceof Capability ) {
+                Capability cap = (Capability) item;
+                FObject capDataObject = null;
+                if ( cap.getOf() != null ){
+                  try {
+                    for ( var capability : capabilities ) {
```

Comment:
capabilities should be a map with a single lookup instead of looping through an array.